src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -258,17 +258,22 @@ private async ValueTask WriteHeadersAsync(HttpHeaders headers, string? cookiesFr
         {
             Debug.Assert(_currentRequest != null);
 
-            if (headers.HeaderStore != null)
+            if (headers.GetEntries() is HeaderEntry[] entries)
             {
-                foreach (KeyValuePair<HeaderDescriptor, object> header in headers.HeaderStore)
+                foreach (HeaderEntry header in entries)
                 {
-                    if (header.Key.KnownHeader != null)
+                    if (header.Key.Descriptor is null)
```

Comment:
Not all elements of the `HeaderEntry[]` may be populated. We don't store the count, so a default (null) descriptor represents the end of the collection.

Reply:
That wasn't obvious to me. Seems like we should at least add a comment.

But I wonder if it would be better to encapsulate this in a custom enumerator, since we do this same logic in multiple places.

Reply:
👍 on adding comments where this happens.

There are only a handful of internal sites using this in such a way. I'm not sure implementing an enumerator and paying the overhead for it is worth it here.